src/engine/ExecuteUpdate.cpp

```diff
@@ -46,61 +46,101 @@ std::pair<std::vector<ExecuteUpdate::TransformedTriple>, LocalVocab>
+        // Ignore the default graph (std::monostate) here, because it is always
+        // added.
+        if (const auto* iri = std::get_if<TripleComponent::Iri>(&g)) {
+          lookupItems.insert(TripleComponent(*iri));
+        }
+      };
+  for (const auto& triple : triples) {
+    addTCToLookup(triple.s_);
+    addTCToLookup(triple.p_);
+    addTCToLookup(triple.o_);
+    addGraphToLookup(triple.g_);
+  }
+  // Always adding the default graph is no overhead and we need it for
+  // `defaultGraphIri`.
+  lookupItems.insert(TripleComponent(
+      ad_utility::triple_component::Iri::fromIriref(DEFAULT_GRAPH_IRI)));
+  std::vector lookupVec(lookupItems.begin(), lookupItems.end());
+  ql::ranges::sort(
+      lookupVec, vocab.getCaseComparator(), [](const TripleComponent& tc) {
+        AD_CORRECTNESS_CHECK(tc.isLiteral() || tc.isIri());
+        return tc.isLiteral() ? tc.getLiteral().toStringRepresentation()
+                              : tc.getIri().toStringRepresentation();
+      });
   // This LocalVocab only contains IDs that are related to the
   // template. Most of the IDs will be added to the DeltaTriples' LocalVocab. An
   // ID will only not be added if it belongs to a Quad with a variable that has
   // no solutions.
   LocalVocab localVocab{};
-
-  auto transformSparqlTripleComponent =
-      [&vocab, &localVocab, &encodedIriManager,
-       &variableColumns](TripleComponent component) -> IdOrVariableIndex {
-    if (component.isVariable()) {
-      AD_CORRECTNESS_CHECK(variableColumns.contains(component.getVariable()));
-      return variableColumns.at(component.getVariable()).columnIndex_;
-    } else {
-      return std::move(component).toValueId(vocab, localVocab,
-                                            encodedIriManager);
+  ad_utility::HashMap<TripleComponent, Id> lookupMap;
+  for (const auto& tc : lookupVec) {
+    // Explicit copy because `toValueId` is `&&`.
+    lookupMap.emplace(tc, TripleComponent(tc).toValueId(vocab, localVocab,
+                                                        encodedIriManager));
+  }
+  auto lookupTc = [&encodedIriManager, &lookupMap, &variableColumns](
+                      const TripleComponent& tc) -> IdOrVariableIndex {
+    // Everything except for Literals and Iris can directly be converted to Ids.
+    if (tc.isVariable()) {
+      AD_CORRECTNESS_CHECK(variableColumns.contains(tc.getVariable()));
+      return variableColumns.at(tc.getVariable()).columnIndex_;
     }
+    if (!(tc.isLiteral() || tc.isIri())) {
+      return tc.toValueIdIfNotString(&encodedIriManager).value();
+    }
+    auto found = lookupMap.find(tc);
+    AD_CORRECTNESS_CHECK(found != lookupMap.end());
+    return found->second;
   };
-  Id defaultGraphIri = [&transformSparqlTripleComponent] {
-    IdOrVariableIndex defaultGraph = transformSparqlTripleComponent(
+  Id defaultGraphIri = [&lookupTc] {
+    const IdOrVariableIndex defaultGraph = lookupTc(
```

Comment:
Did you mean to make this const just because, or did you mean to make this a const ref?

Reply:
Yes it's "just because".